libs/guid/include/guid/oguid.h

```diff
@@ -0,0 +1,217 @@
+/// <param name="noBraces">if we want no braces (default false)</param>
+/// <returns>void</returns>
+/// <remarks></remarks>
+inline void GUIDToString(const GUID& guid, std::wstring& strOut, bool noBraces = false) noexcept
+{
+  auto str = ToString(guid, noBraces);
+  std::swap(str, strOut);
+}
+
+/// <summary>
+/// convert guid string into GUID structure
+/// </summary>
+/// <param name="str">input guid string</param>
+/// <param name="guid">output GUID</param>
+/// <returns>void</returns>
+/// <remarks></remarks>
+/// <exception cref="et::OutOfMemory">memory failure in IIDFromString API</exception>
+/// <exception cref="et::InvalidArgument">bad guid string</exception>
+/// <exception cref="HRESULT">unexpected failure in IIDFromString API</exception>
+inline void StringToGUID(const std::wstring& str, GUID& guid)
+{
+  std::wstring localStr = str;
+  // ensure that the localStr is well formed i.e. GUID is enclosed in {}
+  if (localStr[0] != L'{' && localStr[localStr.length() - 1] != L'}')
+  {
+    std::wstringstream formattedGuidString;
+    formattedGuidString << L"{" << localStr << L"}";
+    localStr = formattedGuidString.str();
+  }
+  HRESULT hr = ::IIDFromString(&localStr[0], &guid);
+  if (hr == E_INVALIDARG)
+  {
+    throw std::invalid_argument("Invalid GUID string");
+  }
+  else if (hr == E_OUTOFMEMORY)
+  {
+    throw std::runtime_error("memory failure converting string to guid");
+  }
+  else if (hr != S_OK)
+  {
+    throw std::exception("Unexpected failure converting string to guid");
+  }
+}
+
+inline GUID StringToGUID(const std::wstring& str)
+{
+  GUID guid{};
+  StringToGUID(str, guid);
+  return guid;
+}
+
+inline GUID StringToGUID(const std::string_view& guidStr)
+{
+  std::wstring wStr(guidStr.data(), guidStr.data() + guidStr.size());
+  return StringToGUID(wStr);
+}
+
+// a const char* member of the overload set is required as long as we provide both basic_string<> and
+// basic_string_view<> due to ambiguous implict construction.  In the future, we should consider if
+// its worth it to collapse this entire space down to basic_string_view<> as the only input parameter.
+// basic_string<> knows how to implicitly convert to basic_string_view<> and basic_string_view<> has
+// a ctor for the CharT* types so the basic_string_view<> type should, in theory, be All We Ever Need (tm)
+inline GUID StringToGUID(const char* guidStr)
+{
+  return StringToGUID(std::string_view(guidStr));
+}
+
+// The wc16* workarounds for wstring_view => wstring don't exist yet
+#if !MS_TARGET_ANDROID && !MS_TARGET_APPLE
+
+inline GUID StringToGUID(const std::wstring_view& guidStr)
+{
+  std::wstring wStr(guidStr);
+  return StringToGUID(wStr);
+}
+
+inline GUID StringToGUID(const wchar_t* guidStr)
+{
+  return StringToGUID(std::wstring_view(guidStr));
+}
+
+#endif
+
+inline std::optional<GUID> TryStringToGUID(const std::wstring& str) noexcept
+{
+  try
+  {
+    return StringToGUID(str);
+  }
+  catch (const std::exception&)

```

Comment:
>catch (const std::exception&) [](http://example.com/codeflow?start=2&length=29)

First we convert error codes to exceptions, and then exceptions to error state? Could we avoid the exceptions all together? They are extremely expensive for such type of operations.